lib/registry/src/publish.rs

```diff
@@ -112,11 +179,19 @@ pub fn try_chunked_uploading(
             package.version
         )
     })?;
+    Ok(url)
+}
 
-    let signed_url = url.url;
-    let url = url::Url::parse(&signed_url).unwrap();
+fn upload_package(
+    signed_url: &str,
+    archive_path: &PathBuf,
+    archived_data_size: u64,
+    timeout: Duration,
+) -> Result<(), anyhow::Error> {
+    let url = url::Url::parse(signed_url).unwrap();
     let client = reqwest::blocking::Client::builder()
         .default_headers(reqwest::header::HeaderMap::default())
+        .timeout(timeout)
```

Comment:
Note that doing it this way means we'll *always* apply the timeout, not just when the user does `--wait`. This should be fine because the non-waiting publish returns pretty quickly, but I just thought I'd point it out.